FlexMeter.c

```diff
@@ -0,0 +1,216 @@
+/*
+htop - FlexMeter.c
+(C) 2024 Stoyan Bogdanov
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <time.h>
+#include <pwd.h>
+#include "FlexMeter.h"
+#include "Object.h"
+#include "config.h"
+#include "CRT.h"
+
+#define FLEX_CFG_FOLDER ".config/htop/FlexMeter"
+
+typedef struct {
+   char* name;
+   char* command;
+   char* type;
+   char* caption;
+   char* uiName;
+} _flex_meter;
+
+_flex_meter meter_list[30];
+
+int meter_list_idx = 0;
+static int meters_count = 0;
+
+static const int DateMeter_attributes[] = {
+   FLEX
+};
+
+MeterClass* FlexMeter_class = NULL;
+
+int check_for_meters(void);
+
+static int parse_input(char* line)
+{
+   if (!strncmp(line, "name=", 5)) {
+      meter_list[meters_count].uiName = xStrdup(line + 5);
+   } else if (!strncmp(line, "command=", 7)) {
+      meter_list[meters_count].command = xStrdup(line + 8);
+      // DRY run command if it return err does not list the metter
+      char buff[256] = {0};
+      int ret = -1;
+      memset(buff, 0, 256);
+      FILE* fd = popen(meter_list[meters_count].command, "r");
+      if (fd) {
+         if (fgets(buff, 256, fd) == NULL) {
+            ret = pclose(fd);
+         } else {
+            ret = pclose(fd);
+         }
+      }
+
+      if (ret != 0) {
+         free(meter_list[meters_count].command);
+         meter_list[meters_count].command = NULL;
+      }
+   }
+   else if (!strncmp(line, "caption=", 7)) {
+      meter_list[meters_count].caption = xStrdup(line + 8);
+   } else if (!strncmp(line, "type=", 5)) {
+      meter_list[meters_count].type = xStrdup(line + 6);
+   } else {
+      return -1;
+   }
+
+   return 0;
+}
+
+static int load_config(char* file)
+{
+   int ret = -1;
+   char* buff;
+   FILE* fp = fopen( file, "r");
+
+   if (fp != NULL) {
+      while (1) {
+         buff = String_readLine(fp);
+         if (buff != NULL) {
+            ret = parse_input(buff);
+            if (ret) {
+               break;
+            }
+         }
+         else {
+            break;
+         }
+      }
+   }
+
+   fclose(fp);
+   return ret;
+}
+
+int check_for_meters(void)
+{
+   char path[400]; // full path
+   DIR* d;
+   struct dirent* dir;
+   struct passwd* pw = getpwuid(getuid());
+   const char* homedir = pw->pw_dir;
```

Comment:
Note the comment by @Explorer09 regarding the `XDG_HOME_DIR` variable and fallbacks …

Reply:
`getpwuid` has some bad side effects (calls to libpam and potentially causes network traffic; thus can potentially hang). Furthermore, when building htop with `--enable-static` this may cause issues, thus call sites should be minimized.

Not to mention the call can fail and is lacking error checks.

Please compare with `Settings_new` in `Settings.c`. Possibly, we should refactor this code block to be a separate function. @Explorer09?